crates/hir-ty/src/method_resolution.rs

```diff
@@ -1079,6 +1081,14 @@ fn iterate_method_candidates_by_receiver(
     table.run_in_snapshot(|table| {
         let mut autoderef = autoderef::Autoderef::new(table, receiver_ty.clone(), true);
         while let Some((self_ty, _)) = autoderef.next() {
+            if !is_receiver_ty_unknown {
+                let canonical_self_ty = autoderef.table.canonicalize(self_ty.clone());
+                if canonical_self_ty.value.is_general_var(Interner, &canonical_self_ty.binders) {
```

Comment:
I think here that can just be `self_ty.inference_var(Interner).is_some()` then?

Reply:
Unfortunately, when I replace my `canonical_self_ty.value.is_general_var(Interner, &canonical_self_ty.binders)` with your suggested `self_ty.inference_var(Interner).is_some()` here, the following tests (which were passing before) now fail:

```
test consteval::tests::trait_method ... FAILED
test tests::traits::trait_method_with_scalar_fallback ... FAILED
test tests::traits::variable_kinds_2 ... FAILED
```

Reply:
Maybe try `autoderef.table.resolve_ty_shallow(&self_ty).inference_var(Interner).is_some()` then. I don't think canonicalizing makes sense here.

Reply:
Using `autoderef.table.resolve_ty_shallow(&self_ty).inference_var(Interner).is_some()` the same 3 tests still fail:

```
failures:

---- consteval::tests::trait_method stdout ----
thread 'consteval::tests::trait_method' panicked at crates/hir-ty/src/consteval/tests.rs:583:5:
Error in evaluating goal: UnresolvedMethod("f")

note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- tests::traits::trait_method_with_scalar_fallback stdout ----
thread 'tests::traits::trait_method_with_scalar_fallback' panicked at crates/hir-ty/src/tests/traits.rs:4328:5:
assertion `left == right` failed: type annotation differs at 195..196
  left: "{unknown}"
 right: "i32"

---- tests::traits::variable_kinds_2 stdout ----
thread 'tests::traits::variable_kinds_2' panicked at crates/hir-ty/src/tests/traits.rs:3244:5:
assertion `left == right` failed: type annotation differs at 99..106
  left: "{unknown}"
 right: "u128"


failures:
    consteval::tests::trait_method
    tests::traits::trait_method_with_scalar_fallback
    tests::traits::variable_kinds_2
```

Reply:
Ok, what you actually need is `matches!(self_ty.kind(Interner), TyKind::InferenceVar(_, TyVariableKind::General))`, because integer or float variables can actually have methods.

Reply:
Yes that works. I will make the change as you suggest and update the pull request.